ALSpringBackend/src/main/java/com/ardaslegends/data/service/Pathfinder.java

```diff
@@ -0,0 +1,156 @@
+      Set<Region> neighbourRegions = currentNode.getNeighboringRegions();
+      for (Region neighbourRegion : neighbourRegions) {
+        //add node to queue if not already visited
+        if (
+          !visitedNodes.contains(neighbourRegion.getId()) &&
+          !nodesToVisitQueue.contains(neighbourRegion)
+        ) {
+          nodesToVisitQueue.add(neighbourRegion);
+        }
+
+        // CALCULATE COST
+        int thisDist = 0;
+
+        // Check if we can move to that region as an army
+        if (player.getRpChar().getBoundTo() != null || isLeaderMove) {
+          if (player.getFaction().getAllies().stream().noneMatch(faction ->
+                  neighbourRegion.getClaimedBy().contains(faction))
+                  && !neighbourRegion.getClaimedBy().contains(player.getFaction())
+                  && !neighbourRegion.getClaimedBy().isEmpty()) {
+            thisDist = 1000;
+          }
+        }
+
+        // Check if we are embarking or disembarking
+        if (currentNode.getRegionType() != RegionType.SEA && neighbourRegion.getRegionType() == RegionType.SEA) {
+          boolean canEmbark = false;
+          for (ClaimBuild claimbuild : currentNode.getClaimBuilds()) {
+            if (claimbuild.getSpecialBuildings().contains(SpecialBuilding.HARBOUR)) {
+              thisDist++;
+              canEmbark = true;
+              break;
+            }
+          }
+          // It's a code smell I know. Had to find a quick fix.
+          // Should be part of the next code review.
+          if (!canEmbark) {
+            thisDist = 1000;
+          }
+        } else if (currentNode.getRegionType() == RegionType.SEA && neighbourRegion.getRegionType() != RegionType.SEA) {
+          thisDist += dist + neighbourRegion.getCost() + 1;
+        } else {
+          thisDist += dist + neighbourRegion.getCost();
+          // Check if there is no army bound to character
+          if (player.getRpChar().getBoundTo() == null && !isLeaderMove) {
+            thisDist /= 2;
+          }
+        }
+
+
+        //if we already have a distance to neighbourRegion, compare with this distance
+        if (prevNodes.containsKey(neighbourRegion)) {
+          //get the recorded smallest distance
+          final int altDist = smallestWeights.get(neighbourRegion);
+
+          //if this distance is better, update the smallest distance + prev node
+          if (thisDist < altDist) {
+            prevNodes.put(neighbourRegion, currentNode);
+            smallestWeights.put(neighbourRegion, thisDist);
+          }
+        } else {
+          //if there is no distance recoded yet, add now
+          prevNodes.put(neighbourRegion, currentNode);
+          smallestWeights.put(neighbourRegion, thisDist);
+        }
+      }
+
+      //mark that we've visited this node
+      visitedNodes.add(currentNode.getId());
+
+      //exit if done
+      if (nodesToVisitQueue.size() == 0) {
+        break;

```

Comment:
Question. If I have read the code correctly, we are ending the algorithm as soon as we went through every single region on the map right?

Reply:
Yep